File: src/tron-address.ts
import { sha256 } from '@noble/hashes/sha2.js';
import { TxCompilerError } from './errors.js';

const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
const HEX_RE = /^[0-9a-fA-F]+$/;
const TRON_ADDRESS_HEX_RE = /^(0x)?41[0-9a-fA-F]{40}$/;
const TRON_ADDRESS_BYTES = 21;
const BASE58CHECK_BYTES = 25;

export function tronAddressToBytes(address: string): Uint8Array {
  if (TRON_ADDRESS_HEX_RE.test(address)) {
    const hex = address.startsWith('0x') ? address.slice(2) : address;
    return hexStringToBytes(hex);
  }

  if (!address.startsWith('T')) {
    throw invalidTronAddress(address, `Invalid Tron address format: ${address}`);
  }

  return base58CheckDecode(address);
}

export function hexToTronAddress(hex20: string): string {
  const clean = hex20.startsWith('0x') ? hex20.slice(2) : hex20;

  if (clean.length !== 40 || !HEX_RE.test(clean)) {
    throw new TxCompilerError('INVALID_ADDRESS', 'Expected 20-byte hex address', {
      address: hex20,
    });
  }

  const payload = new Uint8Array(TRON_ADDRESS_BYTES);
  payload[0] = 0x41;

  for (let i = 0; i < 20; i++) {
    payload[i + 1] = Number.parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }

  const checksum = sha256(sha256(payload)).slice(0, 4);
  const full = new Uint8Array(BASE58CHECK_BYTES);
  full.set(payload);
  full.set(checksum, TRON_ADDRESS_BYTES);

  return base58Encode(full);
}

export function isValidTronAddress(address: string): boolean {
  try {
    tronAddressToBytes(address);
    return true;
  } catch (error) {
    if (error instanceof TxCompilerError && error.code === 'INVALID_ADDRESS') {
      return false;
    }

    throw error;
  }
}

function base58CheckDecode(encoded: string): Uint8Array {
  const raw = base58Decode(encoded);

  if (raw.length !== BASE58CHECK_BYTES) {
    throw invalidTronAddress(encoded, 'Base58check payload must be 25 bytes');
  }

  const payload = raw.slice(0, TRON_ADDRESS_BYTES);
  const checksum = raw.slice(TRON_ADDRESS_BYTES);
  const expected = sha256(sha256(payload)).slice(0, 4);

  for (let i = 0; i < expected.length; i++) {
    if (checksum[i] !== expected[i]) {
      throw invalidTronAddress(encoded, 'Invalid base58check checksum');
    }
  }

  if (payload[0] !== 0x41) {
    throw invalidTronAddress(encoded, 'Tron base58 payload must start with 0x41');
  }

  return payload;
}

function base58Decode(encoded: string): Uint8Array {
  let num = 0n;

  for (const char of encoded) {
    const index = BASE58_ALPHABET.indexOf(char);
    if (index === -1) {
      throw invalidTronAddress(encoded, `Invalid base58 character: ${char}`);
    }

    num = num * 58n + BigInt(index);
  }

  let bytes: Uint8Array;
  if (num === 0n) {
    bytes = new Uint8Array(0);
  } else {
    const hex = num.toString(16);
    const paddedHex = hex.length % 2 === 0 ? hex : '0' + hex;
    bytes = hexStringToBytes(paddedHex);
  }

  let leadingZeros = 0;
  for (const char of encoded) {
    if (char !== '1') break;
    leadingZeros++;
  }

  if (leadingZeros === 0) {
    return bytes;
  }

  const result = new Uint8Array(leadingZeros + bytes.length);
  result.set(bytes, leadingZeros);
  return result;
}

function base58Encode(bytes: Uint8Array): string {
  let leadingZeros = 0;
  for (const byte of bytes) {
    if (byte !== 0) break;
    leadingZeros++;
  }

  let num = 0n;
  for (const byte of bytes) {
    num = num * 256n + BigInt(byte);
  }

  let encoded = '';
  while (num > 0n) {
    const remainder = Number(num % 58n);
    num = num / 58n;
    encoded = BASE58_ALPHABET[remainder] + encoded;
  }

  return '1'.repeat(leadingZeros) + encoded;
}

function hexStringToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);

  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }

  return bytes;
}

function invalidTronAddress(address: string, message: string): TxCompilerError {
  return new TxCompilerError('INVALID_ADDRESS', message, { address });
}

File: src/tron_address.rs
//! Tron address decoding and shape validation.
//!
//! Accepts either the 41-prefixed hex form (`41…`, optionally `0x` prefixed)
//! or the `T…` base58check encoding used by public tooling. Returns the
//! 21-byte raw address (leading `0x41` byte + 20-byte public-key hash).
//!
//! Every return path is guaranteed to produce exactly 21 bytes with a
//! leading `0x41` — callers that slice `&bytes[1..21]` depend on this
//! invariant.

use crate::error::{TxCompilerError, TxCompilerErrorCode};

/// Expected byte length of a decoded Tron address: `0x41` prefix +
/// 20-byte public-key hash.
const TRON_ADDRESS_LEN: usize = 21;

/// Tron address prefix byte.
const TRON_ADDRESS_PREFIX: u8 = 0x41;

/// Decode a Tron address into its raw 21-byte representation.
///
/// On success the returned `Vec<u8>` is always exactly 21 bytes long
/// (the `0x41` prefix byte plus the 20-byte public-key hash). Malformed
/// inputs (wrong prefix, wrong length, invalid base58check, junk hex)
/// return an `InvalidAddress` error.
pub fn tron_address_to_bytes(address: &str) -> Result<Vec<u8>, TxCompilerError> {
    if let Some(stripped) = strip_hex_prefix(address) {
        let bytes = hex::decode(stripped).map_err(|_| invalid_address(address))?;
        enforce_shape(&bytes, address)?;
        return Ok(bytes);
    }

    // Tron base58check addresses are always exactly 34 characters; the
    // decoder would tolerate shorter input and then `enforce_shape` would
    // reject the result, but tightening here shrinks the attack surface
    // of `bs58::decode`.
    if address.starts_with('T') && address.len() == 34 {
        let bytes = bs58::decode(address)
            .with_check(None)
            .into_vec()
            .map_err(|_| invalid_address(address))?;
        enforce_shape(&bytes, address)?;
        return Ok(bytes);
    }

    Err(invalid_address(address))
}

/// Enforce that a decoded byte sequence has the canonical 21-byte
/// Tron address shape with a `0x41` prefix. Defence-in-depth against
/// upstream parser changes (e.g. a different hex-prefix strategy in
/// [`strip_hex_prefix`] that ever admits a non-`41` leading byte).
fn enforce_shape(bytes: &[u8], address: &str) -> Result<(), TxCompilerError> {
    if bytes.len() != TRON_ADDRESS_LEN || bytes[0] != TRON_ADDRESS_PREFIX {
        return Err(invalid_address(address));
    }
    Ok(())
}

/// Return the hex payload if `address` matches `^(0x)?41[0-9a-fA-F]{40}$`.
fn strip_hex_prefix(address: &str) -> Option<&str> {
    let unprefixed = address.strip_prefix("0x").unwrap_or(address);
    if unprefixed.len() != 42 {
        return None;
    }
    let bytes = unprefixed.as_bytes();
    if bytes[0] != b'4' || bytes[1] != b'1' {
        return None;
    }
    if unprefixed[2..].bytes().all(|b| b.is_ascii_hexdigit()) {
        Some(unprefixed)
    } else {
        None
    }
}

fn invalid_address(address: &str) -> TxCompilerError {
    // Keep the raw address out of the human-readable message — it lives
    // in `details` only, where callers can choose whether to redact.
    TxCompilerError::with_details(
        TxCompilerErrorCode::InvalidAddress,
        "Invalid Tron address format",
        serde_json::json!({ "address": address }),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decodes_hex_form() {
        let bytes = tron_address_to_bytes("41d8da6bf26964af9d7eed9e03e53415d37aa96045").unwrap();
        assert_eq!(bytes.len(), 21);
        assert_eq!(bytes[0], 0x41);
    }

    #[test]
    fn decodes_hex_form_with_0x_prefix() {
        let bytes = tron_address_to_bytes("0x41d8da6bf26964af9d7eed9e03e53415d37aa96045").unwrap();
        assert_eq!(bytes.len(), 21);
        assert_eq!(bytes[0], 0x41);
    }

    #[test]
    fn decodes_t_prefix_base58check() {
        // Known-valid Tron address.
        let bytes = tron_address_to_bytes("TLsV52sRDL79HXGGm9yzwKibb6BeruhUzy").unwrap();
        assert_eq!(bytes.len(), 21);
        assert_eq!(bytes[0], 0x41);
    }

    #[test]
    fn rejects_short_hex() {
        let err = tron_address_to_bytes("41ab").unwrap_err();
        assert_eq!(err.code, TxCompilerErrorCode::InvalidAddress);
    }

    #[test]
    fn returned_bytes_always_start_with_0x41() {
        // Defence-in-depth: verify the 0x41 prefix is enforced on every
        // successful return path so the `[1..21]` slice in
        // `build_trc20_transfer_data` never reads an unexpected byte.
        for input in [
            "41d8da6bf26964af9d7eed9e03e53415d37aa96045",
            "0x41d8da6bf26964af9d7eed9e03e53415d37aa96045",
            "TLsV52sRDL79HXGGm9yzwKibb6BeruhUzy",
        ] {
            let bytes = tron_address_to_bytes(input).unwrap_or_else(|e| {
                panic!("expected {input} to decode: {e}");
            });
            assert_eq!(bytes.len(), 21, "{input}: wrong byte length");
            assert_eq!(bytes[0], 0x41, "{input}: missing 0x41 prefix");
        }
    }

    #[test]
    fn rejects_non_tron_prefix_hex() {
        let err = tron_address_to_bytes("d8da6bf26964af9d7eed9e03e53415d37aa96045").unwrap_err();
        assert_eq!(err.code, TxCompilerErrorCode::InvalidAddress);
    }

    #[test]
    fn rejects_invalid_base58_checksum() {
        let err = tron_address_to_bytes("TLsV52sRDL79HXGGm9yzwKibb6BeruhUzz").unwrap_err();
        assert_eq!(err.code, TxCompilerErrorCode::InvalidAddress);
    }
}
